gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java

```diff
@@ -594,7 +594,7 @@ public static Builder builder(String name) {
   }
 
   @SuppressWarnings("unchecked")
-  private <T extends ContextAwareMetric> T getOrCreate(String name, ContextAwareMetricFactory<T> factory) {
+  private synchronized <T extends ContextAwareMetric> T getOrCreate(String name, ContextAwareMetricFactory<T> factory) {
```

Comment:
If you do synchronized on this method, you should also do that on methods `register` and `remove`, which are also manipulating the map.